In [3]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import log_loss
from matplotlib import pyplot as plt
import math
import numpy as np


C:\Users\brean\AppData\Local\Temp\ipykernel_35444\2876505151.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


# Actual Solution

In [46]:
df = pd.read_csv("./CollegePlacement.csv")
df['Placement'] = pd.factorize(df['Placement'])[0]
df['Internship_Experience'] = pd.factorize(df['Internship_Experience'])[0]


X = df.drop(["Placement", "College_ID"], axis = 1)

Y = df["Placement"]


logreg = LogisticRegression(random_state=16,  max_iter=1000, penalty=None)

model = logreg.fit(X, Y)
coef_df = {col : coef for col, coef in zip(list(X.columns), model.coef_[0])}
# print(coef_df)
# print(X.columns)
print(f"coefficent: {coef_df}\nintercept: {model.intercept_}")

coefficent: {'IQ': 0.1088480546249706, 'Prev_Sem_Result': 0.20739706715685027, 'CGPA': 1.0503634055233622, 'Academic_Performance': -0.006580066792577768, 'Internship_Experience': 0.03812121633450414, 'Extra_Curricular_Score': -0.01601374794054621, 'Communication_Skills': 0.6480372982054753, 'Projects_Completed': 0.6793231551407346}
intercept: [-29.18952604]


# Numerical Solution

$ g(B^Tx)= \frac{1}{1+e^{-(B^Tx)}}$

$B^Tx=B_0x_0+B_1x_1+...+B_nx_n$

$x_0=1$


$$
y = \begin{cases}
    \text{1 if} & B^Tx\text{ ≥ 0} \\
    \text{0 if} & B^Tx\text{< 0} \\
\end{cases}
$$

In [5]:
df = pd.read_csv("./CollegePlacement.csv")
# df = pd.read_csv("./college_partial.csv")

df['Placement'] = pd.factorize(df['Placement'])[0]
df['Internship_Experience'] = pd.factorize(df['Internship_Experience'])[0]
df

College_ID   IQ  Prev_Sem_Result  CGPA  Academic_Performance  \
0       CLG0030  107             6.61  6.28                     8   
1       CLG0061   97             5.52  5.37                     8   
2       CLG0036  109             5.36  5.83                     9   
3       CLG0055  122             5.47  5.75                     6   
4       CLG0004   96             7.91  7.69                     7   
...         ...  ...              ...   ...                   ...   
9995    CLG0021  119             8.41  8.29                     4   
9996    CLG0098   70             9.25  9.34                     7   
9997    CLG0066   89             6.08  6.25                     3   
9998    CLG0045  107             8.77  8.92                     3   
9999    CLG0060  109             9.41  9.77                     8   

      Internship_Experience  Extra_Curricular_Score  Communication_Skills  \
0                         0                       8                     8   
1                         0                       7                     8   
2                         0                       3                     1   
3                         1                       1                     6   
4                         0                       8                    10   
...                     ...                     ...                   ...   
9995                      0                       1                     8   
9996                      0                       0                     7   
9997                      1                       3                     9   
9998                      0                       7                     5   
9999                      0                       3                     5   

      Projects_Completed  Placement  
0                      4          0  
1                      0          0  
2                      1          0  
3                      1          0  
4                      2          0  
...                  ...        ...  
9995                   0          1  
9996                   2          0  
9997                   5          0  
9998                   1          0  
9999                   5          0  

[10000 rows x 10 columns]

X contains our coefficents (B)

In [50]:
X = df.drop(["Placement", "College_ID"], axis = 1)
X.insert(0, "intercept", [1] * X.shape[0])
Y = df["Placement"].to_numpy()

X

intercept   IQ  Prev_Sem_Result  CGPA  Academic_Performance  \
0             1  107             6.61  6.28                     8   
1             1   97             5.52  5.37                     8   
2             1  109             5.36  5.83                     9   
3             1  122             5.47  5.75                     6   
4             1   96             7.91  7.69                     7   
...         ...  ...              ...   ...                   ...   
9995          1  119             8.41  8.29                     4   
9996          1   70             9.25  9.34                     7   
9997          1   89             6.08  6.25                     3   
9998          1  107             8.77  8.92                     3   
9999          1  109             9.41  9.77                     8   

      Internship_Experience  Extra_Curricular_Score  Communication_Skills  \
0                         0                       8                     8   
1                         0                       7                     8   
2                         0                       3                     1   
3                         1                       1                     6   
4                         0                       8                    10   
...                     ...                     ...                   ...   
9995                      0                       1                     8   
9996                      0                       0                     7   
9997                      1                       3                     9   
9998                      0                       7                     5   
9999                      0                       3                     5   

      Projects_Completed  
0                      4  
1                      0  
2                      1  
3                      1  
4                      2  
...                  ...  
9995                   0  
9996                   2  
9997                   5  
9998                   1  
9999                   5  

[10000 rows x 9 columns]

$ g(B^Tx)= \frac{1}{1+e^{-(B^Tx)}}$

$B^Tx=B_0x_0+B_1x_1+...+B_nx_n$

$x_0=1$


$$
y = \begin{cases}
    \text{1 if} & B^Tx\text{ ≥ 0} \\
    \text{0 if} & B^Tx\text{< 0} \\
\end{cases}
$$

## Cost Function

$i$ is a row $j$ is a column

$ J(B) = \frac{1}{m} \sum [y_i ln(\frac{1}{1+e^{-(B^Tx)}}) + (1-y_i)ln(\frac{1}{1+e^{-(B^Tx)}})]$

$ B_j := B_j-α(\frac{∂J(B)}{∂B_j}) $

$ \frac{∂J(B)}{∂B_j} = \frac{1}{m} \sum [(\frac{1}{1+e^{-(B^Tx)}}) - y_i]x_{j,i}$

In [38]:
LEARNING_RATE = 0.1
EPOCHS = 100


cols = list(X.columns)
# cols[0] = "intercept"
array_x = X.to_numpy()
B_array = np.zeros( X.shape[1]) 


def compute_sigmoid(z):
    return np.where(
            z >= 0, # condition
            1 / (1 + np.exp(-z)), # For positive values
            np.exp(z) / (1 + np.exp(z)) # For negative values
    )


for epoch in range(EPOCHS):
    
    
    ## gradient descent
    # z_values = np.sum(array_x * B_array, axis = 1) ## alt: 
    z_values = array_x @ B_array
    sigmoids = compute_sigmoid(z_values)

    # current_b_partials = np.sum( ( array_x * np.array(sigmoids - Y)[:, np.newaxis] / len(Y) ) , axis = 0 )  ## alt: 
    current_b_partials = np.sum( (sigmoids - Y)[:,np.newaxis] * array_x, axis = 0 ) / Y.shape[0]

    ## update weights
    B_array -= LEARNING_RATE * current_b_partials

    ## compute error
    z_values = np.sum(array_x * B_array, axis = 1) ## alt: array_x @ B_array

    sigmoids = compute_sigmoid(z_values)
    # eps=1e-15
    # sigmoids = np.clip(sigmoids, eps, 1 - eps)
    error = sum( ( Y * np.log(sigmoids) ) + ( (1 - Y) * np.log(1 - sigmoids) ) ) * (-1/len(Y))

    vecb = {col: b for col, b in zip(cols, B_array)}
    print(f"------------------------------------------Epoch: {epoch + 1}------------------------------------------")
    print(f"ERROR: {error}\ncoefficents: {vecb}")

# ERROR: 3.871293008423183
# coefficents: [-0.1807298  -0.31260436  0.23073075  0.27732736 -0.59486144 -0.07222909
#  -0.45740054  1.25252526  0.71935345]

# ERROR: 5.495505791990629
# coefficents: {'intercept': -0.1825955850287635, 'IQ': -0.47057961327600084, 'Prev_Sem_Result': 0.2141773620369364, 'CGPA': 0.2607443850492142, 'Academic_Performance': -0.5963416078438631, 'Internship_Experience': -0.07272034483530379, 'Extra_Curricular_Score': -0.45948800449799254, 'Communication_Skills': 1.2159684708626808, 'Projects_Completed': 0.7110840149600344}


------------------------------------------Epoch: 1------------------------------------------
ERROR: 58.47539190103774
coefficents: {'intercept': -0.03341, 'IQ': -3.16326, 'Prev_Sem_Result': -0.23465015, 'CGPA': -0.23406965000000005, 'Academic_Performance': -0.1869, 'Internship_Experience': -0.01335, 'Extra_Curricular_Score': -0.166635, 'Communication_Skills': -0.15095000000000003, 'Projects_Completed': -0.07009}
------------------------------------------Epoch: 2------------------------------------------
ERROR: 24.952664169510292
coefficents: {'intercept': -0.01682, 'IQ': -1.35293, 'Prev_Sem_Result': -0.09251664999999998, 'CGPA': -0.09152035000000006, 'Academic_Performance': -0.09648000000000001, 'Internship_Experience': -0.006880000000000001, 'Extra_Curricular_Score': -0.084725, 'Communication_Skills': -0.023810000000000026, 'Projects_Completed': -0.014510000000000002}
------------------------------------------Epoch: 3------------------------------------------
ERROR: nan
coefficents: {

C:\Users\brean\AppData\Local\Temp\ipykernel_35444\2895310541.py:39: RuntimeWarning: divide by zero encountered in log
  error = sum( ( Y * np.log(sigmoids) ) + ( (1 - Y) * np.log(1 - sigmoids) ) ) * (-1/len(Y))
C:\Users\brean\AppData\Local\Temp\ipykernel_35444\2895310541.py:39: RuntimeWarning: invalid value encountered in multiply
  error = sum( ( Y * np.log(sigmoids) ) + ( (1 - Y) * np.log(1 - sigmoids) ) ) * (-1/len(Y))
C:\Users\brean\AppData\Local\Temp\ipykernel_35444\2895310541.py:14: RuntimeWarning: overflow encountered in exp
  1 / (1 + np.exp(-z)), # For positive values


------------------------------------------Epoch: 63------------------------------------------
ERROR: 4.025906423479486
coefficents: {'intercept': -0.10483056704055771, 'IQ': -0.3418124579956443, 'Prev_Sem_Result': 0.28838219481547533, 'CGPA': 0.31836569305505136, 'Academic_Performance': -0.6819013474923767, 'Internship_Experience': -0.048250357821892714, 'Extra_Curricular_Score': -0.55620607253954, 'Communication_Skills': 1.6137445946190518, 'Projects_Completed': 0.6111284574317061}
------------------------------------------Epoch: 64------------------------------------------
ERROR: nan
coefficents: {'intercept': -0.08826972041317514, 'IQ': 1.4667875297503306, 'Prev_Sem_Result': 0.43029927014635316, 'CGPA': 0.4606986751376385, 'Academic_Performance': -0.5915495150542218, 'Internship_Experience': -0.0417971689290341, 'Extra_Curricular_Score': -0.47434393373421524, 'Communication_Skills': 1.7406030660570964, 'Projects_Completed': 0.6666223580635872}
---------------------------------------

In [52]:
Y = df["Placement"].to_numpy()
array_x = X.to_numpy()

In [62]:
LEARNING_RATE = 0.01
EPOCHS = 10000

b_vector = np.zeros(array_x.shape[1]) ## initilize intercept and coefficents as 0
cols = list(X.columns)


def compute_sigmoid(B: np.ndarray, X: np.ndarray):

    # return 1 / (1 + np.exp(-np.dot(X, B))) ## ALT METHOD: np.dot(X, B) computes the dot product between each row of X and B
    # return 1/(  1 + np.exp( - X @ B )  ) ## X @ B computes the dot product between each row of X and B
    return np.where(
            X @ B >= 0, # condition
            1 / (1 + np.exp(-X @ B)), # For positive values
            np.exp(X @ B) / (1 + np.exp(X @ B)) # For negative values
    )


def compute_cost(B: np.ndarray, X: np.ndarray, Y: np.ndarray):
    num_rows = Y.shape[0]
    sigmoid = compute_sigmoid(B, X)

    return np.sum((Y * np.log(sigmoid)) + (1 - Y) * np.log(sigmoid)) / -num_rows

def partial_derivative(B: np.ndarray, X: np.ndarray, Y: np.ndarray):
    num_rows = Y.shape[0]
    sigmoid = compute_sigmoid(B, X)

    # return (1 / num_rows) * np.dot(X.T, (sigmoid - Y)) ## ALT METHOD
    
    ## (sigmoid - Y)[:,np.newaxis] changes its shape from (n,) to (n, 1)
    return np.sum( (sigmoid - Y)[:,np.newaxis] * X, axis = 0 ) / num_rows


for epoch in range(EPOCHS):

    cost = compute_cost(b_vector, array_x, Y)
    
    b_vector -= LEARNING_RATE * partial_derivative(b_vector, array_x, Y)

    b_dict = {col : b for col, b in zip(cols, b_vector)}

    if epoch % 100 == 0 or epoch+1 == EPOCHS:
        print(f"------------------------------------epoch: {epoch}------------------------------------")
        print(f"\t\t\t\tcost: {cost}\nb_dict: {b_dict}")


print("final values: ")
for key, value in b_dict.items(): print(f"{key}: {round(value,8)}")


------------------------------------epoch: 0------------------------------------
				cost: 0.6931471805599453
b_dict: {'intercept': -0.0033410000000000002, 'IQ': -0.316326, 'Prev_Sem_Result': -0.023465015000000002, 'CGPA': -0.023406965000000005, 'Academic_Performance': -0.018690000000000002, 'Internship_Experience': -0.001335, 'Extra_Curricular_Score': -0.0166635, 'Communication_Skills': -0.015095, 'Projects_Completed': -0.007009}
------------------------------------epoch: 100------------------------------------
				cost: 2.9878329471667078
b_dict: {'intercept': -0.015769791650236058, 'IQ': 0.06864728701853134, 'Prev_Sem_Result': 0.05214419267598205, 'CGPA': 0.05693847091476238, 'Academic_Performance': -0.09502505554374435, 'Internship_Experience': -0.00719451431653022, 'Extra_Curricular_Score': -0.07623755850856173, 'Communication_Skills': 0.2444962286739376, 'Projects_Completed': 0.09786997076842936}
------------------------------------epoch: 200------------------------------------
	

In [45]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as api
import statsmodels

df = pd.read_csv("./CollegePlacement.csv")
# df = pd.read_csv("./college_partial.csv")

df['Placement'] = pd.factorize(df['Placement'])[0]
df['Internship_Experience'] = pd.factorize(df['Internship_Experience'])[0]

df = df.drop(["College_ID"], axis = 1)
# Y = df["Placement"]

df_with_intercept = statsmodels.tools.tools.add_constant(df, prepend=True)
model = api.logit(formula="Placement ~ IQ + Prev_Sem_Result + CGPA + Academic_Performance + Internship_Experience + Extra_Curricular_Score + Communication_Skills + Projects_Completed", data = df_with_intercept)

model.fit().summary()


Optimization terminated successfully.
         Current function value: 0.218626
         Iterations 9


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Placement   No. Observations:                10000
Model:                          Logit   Df Residuals:                     9991
Method:                           MLE   Df Model:                            8
Date:                Mon, 29 Sep 2025   Pseudo R-squ.:                  0.5134
Time:                        10:09:18   Log-Likelihood:                -2186.3
converged:                       True   LL-Null:                       -4493.3
Covariance Type:            nonrobust   LLR p-value:                     0.000
==========================================================================================
                             coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                -29.1898      0.733    -39.843      0.000     -30.626     -27.754
IQ                         0.1088      0.004     31.003      0.000       0.102       0.116
Prev_Sem_Result            0.2072      0.137      1.508      0.131      -0.062       0.477
CGPA                       1.0505      0.137      7.642      0.000       0.781       1.320
Academic_Performance      -0.0066      0.013     -0.494      0.621      -0.033       0.020
Internship_Experience      0.0383      0.079      0.484      0.628      -0.117       0.193
Extra_Curricular_Score    -0.0160      0.012     -1.311      0.190      -0.040       0.008
Communication_Skills       0.6481      0.019     33.287      0.000       0.610       0.686
Projects_Completed         0.6794      0.027     25.044      0.000       0.626       0.733
==========================================================================================
"""

coefficent: {'IQ': 0.1088480546249706, 'Prev_Sem_Result': 0.20739706715685027, 'CGPA': 1.0503634055233622, 'Academic_Performance': -0.006580066792577768, 'Internship_Experience': 0.03812121633450414, 'Extra_Curricular_Score': -0.01601374794054621, 'Communication_Skills': 0.6480372982054753, 'Projects_Completed': 0.6793231551407346}
intercept: [-29.18952604]

In [225]:
np.set_printoptions(threshold=np.inf)
# probs = sigmoid(X_with_intercept @ B)
# print("Predicted probabilities:", probs)


```python
array_x = [[1, 1, 4, 7],
           [1, 2, 5, 8],
           [1, 3, 6, 9]]

B_array = [1, 1, 1, 1] # b0, b1, b2, b3 all equal 1

array_x * B_array = 
[1, 1, 4, 7] * [b0, b1, b2, b3] = [1 * b0, 1 * b1, 4 * b2, 7 * b3]
[1, 2, 5, 8] * [b0, b1, b2, b3] = [1 * b0, 2 * b1, 6 * b2, 8 * b3]
[1, 3, 6, 9] * [b0, b1, b2, b3] = [1 * b0, 3 * b1, 5 * b2, 9 * b3]

## Lets say b0, b1, b2, b3 all equal 1
## This computes B^T*x
## These are by z_values
np.sum(array_x * B_array, axis = 1) = [sum([1 * b0, 1 * b1, 4 * b2, 7 * b3]), 
                                      sum([1 * b0, 2 * b1, 6 * b2, 8 * b3]), 
                                      sum([1 * b0, 3 * b1, 5 * b2, 9 * b3])]  
                                      
                                    = [13, 17, 18]


```